File: Python/DataStructures/BinarySearchTree.py

class TreeNode:
    def __init__(self, value, right, left):
        self.value = value
        self.right = right
        self.left = left
    

class BinarySearchTree:

    def __init__(self):
        self.root = None

    def isEmpty(self):
        return self.root is None
    
    def add(self, value):
        if self.root is None:
            self.root = TreeNode(value, None, None)
            return
        self._addRecursive(value, self.root)

    def _addRecursive(self, value, node):
        if value > node.value:
            if node.right is not None:
                self._addRecursive(value, node.right)
            else:
                node.right = TreeNode(value, None, None)
        elif value < node.value:
            if node.left is not None:
                self._addRecursive(value, node.left)
            else:
                node.left = TreeNode(value, None, None)
            
    def contains(self, value):
        if self.root is None:
            return False
        return self._containsRecursive(value, self.root)
    
    def _containsRecursive(self, value, node):
        if node is None:
            return False
        if value > node.value:
            return self._containsRecursive(value, node.right)
        elif value < node.value:
            return self._containsRecursive(value, node.left)
        else:
            return True
    
    def find(self, value):
        if self.root is None:
            return None
        return self._findRecursive(value, self.root)

    def _findRecursive(self, value, node):
        if node is None:
            return None
        
        if value > node.value:
            return self._findRecursive(value, node.right)
        elif value < node.value:
            return self._findRecursive(value, node.left)
        else:
            return node
    
    def remove(self, value):
        if self.root is None:
            return
        else:
            self.root = self._removeRecursive(value, self.root)
    
    def _removeRecursive(self, value, node):

        if node is None:
            return None
        
        if value > node.value:
            node.right = self._removeRecursive(value, node.right)
        elif value < node.value:
            node.left = self._removeRecursive(value, node.left)
        else:
            if node.left is not None and node.right is not None:
                tmp = node
                leftMost = self._getLeftMost(tmp.right)
                node.value = leftMost.value

                self.root.right = self._removeRecursive(leftMost.value, node.right)

            elif node.left is not None:
                node = node.left
            elif node.right is not None:
                node = node.right
            else:
                node = None
                
        return node
    
    def _getLeftMost(self, node):
        if node is None:
            return None
        elif node.left is None:
            return node
        else:
            return self._getLeftMost(node.left)

    def clear(self):
        self.root = None

    def __str__(self):
        return "[" + self.toString(self.root) + "]"

    def toString(self, node):
        res = ""
        if node is None:
            return res
        
        if node.left is not None:
            res += self.toString(node.left)
        res += str(node.value) + " "
        if node.right is not None:
            res += self.toString(node.right)
        return res

File: TypeScript/DataStructures/BinarySearchTree.ts
class BSTNode
{
    public value: number;
    public right: BSTNode;
    public left: BSTNode;
}

class BinarySearchTree
{
    private root: BSTNode;

    public isEmpty(): boolean
    {
        return this.root == null;
    }   

    public add(value: number)
    {
        if(this.isEmpty())
        {
            this.root = new BSTNode();
            this.root.value = value;
        }
        else
        {
            this.addAux(value, this.root);
        }
    }
    
    // Helper method for add:
    private addAux(value: number, treeNode: BSTNode)
    {
        if(value > treeNode.value)
        {
            if(treeNode.right != null)
            {
                this.addAux(value, treeNode.right);
            }
            else
            {
                treeNode.right = new BSTNode();
                treeNode.right.value = value;
            }
        }
        else if(value < treeNode.value)
        {
            if(treeNode.left != null)
            {
                this.addAux(value, treeNode.left);
            }
            else
            {
                treeNode.left = new BSTNode();
                treeNode.left.value = value;
            }
        }
        // Equal: do nothing. (you could also throw.)
    }

    public contains(value: number): boolean
    {
        if(this.isEmpty()) return false;
        return this.containsAux(value, this.root);
    }

    private containsAux(value: number, treeNode: BSTNode): boolean
    {
        if(treeNode == null) return false;

        if(value > treeNode.value)
        {
            return this.containsAux(value, treeNode.right);
        }
        else if (value < treeNode.value)
        {
            return this.containsAux(value, treeNode.left);
        }
        else
        {
            return true;
        }
    }

    public remove(value: number)
    {
        if(this.root == null) return;
        this.root = this.removeAux(value, this.root);
    }

    private removeAux(value: number, treeNode: BSTNode): BSTNode
    {
        if(treeNode == null)
        {
            return null;
        }

        if(value > treeNode.value)
        {
            treeNode.right = this.removeAux(value, treeNode.right);
        }
        else if(value < treeNode.value)
        {
            treeNode.left = this.removeAux(value, treeNode.left);
        }
        else
        {
            if(treeNode.left != null && treeNode.right != null)
            {
                const tmp = treeNode;
                const leftMost = this.getLeftMost(tmp.right);
                treeNode.value = leftMost.value;
                treeNode.right = this.removeAux(leftMost.value, treeNode.right);
            }
            else if(treeNode.left != null)
            {
                treeNode = treeNode.left;
            }
            else if(treeNode.right != null)
            {
                treeNode = treeNode.right;
            }
            else
            {
                treeNode = null;
            }
        }

        return treeNode;
    }

    private getLeftMost(treeNode: BSTNode): BSTNode
    {
        if(treeNode == null) return null;
        let tmp = treeNode;
        while(tmp.left != null)
        {
            tmp = tmp.left;
        }
        return tmp;
    }
    
}